Reordered size checks in caml_make_vect for better performance

mmottl opened the pull request:
`size <= Max_young_wosize` always implies `size <= Max_wosize`.  Checking for the former first should improve performance, because it is by far the more likely case and we can thus avoid an additional branch in the majority of cases.

mmottl committed:
Reordered and removed redundant size checks in various allocation functions

xavierleroy commented:
Looks good to me.  

Did anyone check whether this pattern occurs elsewhere in array.c or related files?

mshinwell commented:
I found one other occurrence, in `caml_array_gather`.  @mmottl Please re-order the cases in that function too and add a `Changes` entry.  Thanks.

mshinwell commented:
(I checked the whole source tree and didn't find any other places to alter.)

mmottl commented on byterun/alloc.c:
```diff
@@ -38,12 +38,14 @@ CAMLexport value caml_alloc (mlsize_t wosize, tag_t tag)
-    Alloc_small (result, wosize, tag);
-    if (tag < No_scan_tag){
-      for (i = 0; i < wosize; i++) Field (result, i) = Val_unit;
+  if (wosize <= Max_young_wosize){
```
Pushed `wosize == 0` into the `wosize <= Max_young_wosize` branch.  It can thus be avoided when executing the `caml_alloc_shr` branch.

mmottl commented on byterun/alloc.c:
```diff
@@ -138,19 +140,15 @@ CAMLexport value caml_alloc_array(value (*funct)(char const *),
-      caml_modify(&Field(result, n), v);
-    }
-    CAMLreturn (result);
+  result = caml_alloc (nbr, 0);
```
`nbr == 0` means that `caml_alloc` will return `Atom` anyway.  The change makes the atomic case slightly less efficient, because the for-loop check will also happen in addition to the one in `caml_alloc`.  But the atomic case is highly unlikely whereas we can save the extra `nbr == 0` check in the frequent non-atomic case.

mmottl commented on byterun/alloc.c:
```diff
@@ -161,10 +159,11 @@ CAMLprim value caml_alloc_float_array(mlsize_t len)
-    return Atom(0);
-  else if (wosize <= Max_young_wosize){
-    Alloc_small (result, wosize, Double_array_tag);
+  if (wosize <= Max_young_wosize){
```
Pushed `wosize == 0` into `wosize <= Max_young_wosize` branch again.

mmottl commented on byterun/alloc.c:
```diff
@@ -195,8 +194,6 @@ CAMLexport int caml_convert_flag_list(value list, int *flags)
   mlsize_t wosize = Long_val(size);
-
-  if (wosize == 0) return Atom(0);
   return caml_alloc (wosize, 0);
```
`caml_alloc` already handles `wosize == 0`, no need for a shortcut here.

mmottl commented on byterun/alloc.c:
```diff
@@ -195,8 +194,6 @@ CAMLexport int caml_convert_flag_list(value list, int *flags)
 {
   mlsize_t wosize = Long_val(size);
-
-  if (wosize == 0) return Atom(0);
```
`caml_alloc` already handles `wosize == 0`, no need for a shortcut here.

mmottl commented on byterun/alloc.c:
```diff
@@ -211,8 +208,6 @@ CAMLprim value caml_alloc_dummy_function(value size,value arity)
 {
   mlsize_t wosize = Long_val(size) * Double_wosize;
-
-  if (wosize == 0) return Atom(0);
```
`caml_alloc` already handles `wosize == 0`, no need for a shortcut here.

mmottl commented on byterun/array.c:
```diff
@@ -149,12 +149,13 @@ CAMLprim value caml_make_float_vect(value len)
-  if (wosize == 0)
-    return Atom(0);
-  else if (wosize <= Max_young_wosize){
+  if (wosize <= Max_young_wosize){
```
Pushed `wosize == 0` into `wosize <= Max_young_wosize` branch again.

mmottl commented on byterun/array.c:
```diff
@@ -176,27 +177,32 @@ CAMLprim value caml_make_vect(value len, value init)
-    res = caml_alloc(wsize, Double_array_tag);
-    for (i = 0; i < size; i++) {
-      Store_double_field(res, i, d);
+    if (size == 0) {
```
Pushed `wosize == 0` into several branches, avoiding it when possible.

mmottl commented on byterun/intern.c:
```diff
@@ -570,7 +570,7 @@ static void intern_alloc(mlsize_t whsize, mlsize_t num_objects,
   }
   wosize = Wosize_whsize(whsize);
-  if (wosize > Max_wosize || outside_heap) {
+  if (outside_heap || wosize > Max_wosize) {
```
Probably doesn't matter to C-compilers with good optimizations, but anyway.

mmottl commented on byterun/intern.c:
```diff
@@ -585,10 +585,12 @@ static void intern_alloc(mlsize_t whsize, mlsize_t num_objects,
     CAMLassert (intern_block == 0);
   } else {
     /* this is a specialised version of caml_alloc from alloc.c */
```
Pushed `wosize == 0` into `wosize <= Max_young_wosize` branch again.

mmottl commented on byterun/array.c:
```diff
@@ -320,38 +326,47 @@ static value caml_array_gather(intnat num_arrays,
-  if (size == 0) {
-    /* If total size = 0, just return empty array */
-    res = Atom(0);
+  if (isfloat) {
```
Pushed `size == 0` into branches again.

mmottl commented:
I've fixed `caml_array_gather`, too.  I didn't find this particular issue elsewhere, but I noticed that there were several instances of a similar inefficiency: some parts of the code checked for atomic arguments (e.g. via `wosize == 0`).  Some of these checks were redundant (e.g. in `caml_alloc_dummy`) or could be pushed into subsequent branches and eliminated in some of them (e.g. in `caml_alloc_float_array`).

I've added some line comments to explain the changes.  They should be fairly straightforward to verify.

lpw25 commented:
> could be pushed into subsequent branches and eliminated in some of them

Are these cases definitely faster like this? You're swapping a well-predicted branch based on an in-cache value for an increase in code size. It is not immediately clear to me that this will improve performance.

mmottl commented:
@lpw25 Obviously, only benchmarks could really tell and even those only for the tested platforms.  Even changes that in theory should yield monotonic improvements may run into surprises due to modern CPU architecture weirdness.

First of all, the issue you raise only concerns the few patches where there is more than one conditional through which we push checks.  In those cases the marginally increased code size should only matter if parent branches are taken roughly at random.  In practice I'd expect the vast majority of calls going into the same branch as previous calls (e.g. allocating the same size of array in a loop).

I'm not totally attached to the changes that affect multiple branches.  The improvements would likely be small anyway.  A single binary branch, however, should at least in theory benefit if we push a check into the only branch that actually needs it.  In that case code size would stay the same.

lpw25 commented:
> A single binary branch, however, should at least in theory benefit if we push a check into the only branch that actually needs it. In that case code size would stay the same.

Yeah, I have no issue with those changes, they seem obviously faster and have no cost in terms of the readability of the code, even improving it in some cases. I'm less keen on the multiple branch ones since they marginally decrease the readability in exchange for only a possible performance improvement. That said I don't feel strongly either way, just thought it was worth pointing out.

mmottl commented:
@lpw25 That pretty much reflects my take on this.  Lets hear from the OCaml team what their preferences are.  If marginally more readability is more important than marginally more performance, I'll just undo the multi-branch changes.

alainfrisch commented:
Not that I'm against the changes, but I'm genuinely curious about the kind of performance improvement that can be achieved in practice -- I thought it would be largely "in the noise" on "usual" architectures.  Have you observed some improvements on concrete use cases?

mmottl commented:
@alainfrisch Obviously, these changes are micro-optimizations, and I didn't think it would be worthwhile timing them.  Some of them don't change the complexity of the code as in the binary branch cases, some even simplify things, some slightly increase complexity.  I'm fine if we drop the changes changes that increase complexity, but I think that the others are worth it.  One percent here and one percent there for no extra complexity can add up over time.

mshinwell commented:
@mmottl I suggest we remove the changes that increase complexity until such time as someone can demonstrate a benefit.  Then we can merge the others now.  OK?

mmottl commented:
@mshinwell Done.  You may want to proofread changes with `patdiff`.  The full diff to `trunk` is trivial to read that way.

mshinwell commented:
Please add a Changes entry...

mmottl commented:
Changes are updated.  Thanks everyone for reviewing!

mshinwell commented:
@mmottl Sorry, Changes is conflicting again.  Please rebase and I will merge this tomorrow.

mmottl commented:
@mshinwell Do you also want me to squash the changes?

gasche commented:
(Personal opinion: I think it's nice to keep separate changes in separate commits. but then you could squash/fixup the "reverted ..." in the relevant commits to not keep the back-and-forth in the commit history. Having Changes in a separate commit is fine -- there is no commit it should canonically go to. Squashing everything in a single commit would also be acceptable.)

mmottl commented:
It would be good to know what the general policy is wrt. commit hygiene for contributors.  I personally don't have strong feelings for or against commit squashing.  I think unless there is an excessive back and forth it probably doesn't matter.

gasche commented:
I wrote my opinion in [CONTRIBUTING.md: Clean patch series](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#clean-patch-series).

mmottl commented:
@gasche Thanks, it's been a while that I read it.  I have squashed the patches into one and rebased it.
